Feature/db

KPP13 opened the pull request:
Integration between parser and database module.
Crawler logic is capable of inserting the book data into the relational database.
Many tables with different relations: one-to-one, one-to-many, many-to-many are used.

Commit:
Entitiest related with Book
For now I created entities related with books.

czyzu0507 committed:
Adding configuration removed by mistake

czyzu0507 committed:
Reslving problems with Spring data
Because of conflicts beetwen spring and hibernate I cheanged version of those dependencies.
I also introduced DAOs and changed spring configuration to be able to work with spring data.

czyzu0507 committed:
Integration with kamil's branch
I removed his book and category class and changed into my implementation

czyzu0507 committed:
Parser saves data in database and database for test is configured
Now I ask database for every author, category and printHouse.
I think thet it will be changed in the feature.

czyzu0507 committed:
Lots of select commands moved to one
In previous commit I had a version  which ask database for every Author etc.
Now I select all of those into the memory.

czyzu0507 committed:
Cleaning up after rebasing

czyzu0507 committed:
Refactoring and adding javadocs

KPP13 committed:
Fixing & cleaning - just small changes.

czyzu0507 committed:
Pagination
I introduced method for selecting only given part of information about book from database

maciejkocur commented on robot-db/src/main/java/repositories/BookInput.java:
```diff
@@ -0,0 +1,49 @@
+        log.info("staring operations on database");
+        Category category = categoryInput.saveCategory(parsedBook.getCategory());
+        PrintHouse printHouse = printHouseInput.savePrintHouse(parsedBook.getPrintHouse());
+        List<Author> authors = parsedBook.getAuthors().stream().map(authorInput::saveAuthor).collect(Collectors.toList());
```
Consider instead:
List<Author> authors = parsedBook.getAuthors()
                      .stream()
                      .map(authorInput::saveAuthor)
                      .collect(Collectors.toList());

I think it looks better

czyzu0507 replied:
but without diamonds you get Unchecked assignment

maciejkocur commented on robot-db/src/main/java/repositories/BookInput.java:
```diff
@@ -0,0 +1,49 @@
+        List<Author> authors = parsedBook.getAuthors().stream().map(authorInput::saveAuthor).collect(Collectors.toList());
+        Book book = Book.builder().title(parsedBook.getTitle()).category(category).authors(authors).currency(parsedBook.getCurrency())
+                .description(parsedBook.getDescription()).printHouse(printHouse).link(parsedBook.getLink())
+                .newPrice(parsedBook.getNewPrice()).oldPrice(parsedBook.getOldPrice()).year(parsedBook.getYear()).build();
```
same as above

czyzu0507 committed:
Little fix
I noticed thet I had the method name starts with big letter. Shame on me.

czyzu0507 committed:
Refactoring
Now Input clases rerutn values without add them to the map and later get from the map.
I think it will improve performance.

LIttleAncientForestKami commented on pom.xml:
```diff
@@ -25,19 +25,20 @@
         <maven.surefire.plugin.version>2.19</maven.surefire.plugin.version>
         <webapp-runner.version>8.0.33.1</webapp-runner.version>
-        <springframework.version>4.3.2.RELEASE</springframework.version>
+        <springframework.version>4.2.6.RELEASE</springframework.version>
```
Why the update?

LIttleAncientForestKami commented on pom.xml:
```diff
@@ -25,19 +25,20 @@
         <webapp-runner.version>8.0.33.1</webapp-runner.version>
-        <springframework.version>4.3.2.RELEASE</springframework.version>
+        <springframework.version>4.2.6.RELEASE</springframework.version>
+        <hibernate.version>5.1.0.Final</hibernate.version>
```
Why the downgrade?

LIttleAncientForestKami commented on pom.xml:
```diff
@@ -48,6 +49,14 @@
             <scope>test</scope>
         </dependency>
 
+        <!--Spring-Data-->
```
Unnecessary comment. Same with mvnrepository... ones below.

LIttleAncientForestKami commented on robot-db/initdata.script:
```diff
@@ -0,0 +1,4 @@
+CREATE SCHEMA PUBLIC AUTHORIZATION DBA
```
Each script should have a comment explaining why it was created. Presumably together with issue.

LIttleAncientForestKami commented on robot-db/pom.xml:
```diff
@@ -12,5 +11,12 @@
+        <dependency>
+            <groupId>hsqldb</groupId>
+            <artifactId>hsqldb</artifactId>
+            <version>1.8.0.10</version>
```
not in props?

LIttleAncientForestKami commented on robot-db/src/main/java/dao/AuthorDAO.java:
```diff
@@ -0,0 +1,14 @@
+
+
+/**
+ * This class is responsible for managing Authors in database
```
Cut "this class is" please. It's a class Javadoc, you obviously are NOT talking about packages or methods here.

LIttleAncientForestKami commented on robot-db/src/main/java/dao/BookDAO.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.stereotype.Component;
+
+/**
+ * This class is responsible for managing Books in database
```
As above with "this class". Also, Javadocs on public classes should not just be repeating class name. `ClassDao is a DAO for class, so it's responsible for managing a class, as a Data Access Object.` is a weak Javadoc.

LIttleAncientForestKami commented on robot-db/src/main/java/dao/CategoryDAO.java:
```diff
@@ -0,0 +1,14 @@
+
+@Component
+public interface CategoryDAO extends CrudRepository<Category, Long> {
+
```
No extra lines between Javadocs and class it details please. Also, don't write Javadocs for X starting with "This X is responsible". Javadocs are meant for other devs. They know if it's a class, a method, a package or a field. Inform them of important matters, not of all matters.

LIttleAncientForestKami commented on robot-db/src/main/java/dao/PrintHouseDAO.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.data.repository.CrudRepository;
+
+/**
+ * This class is responsible for manging PrintHouses in database
```
This class.
Managing. Not manging. :-)

LIttleAncientForestKami commented on robot-db/src/main/java/dbconfiguration/SpringDBConfiguration.java:
```diff
@@ -2,29 +2,35 @@
     Environment env;
 
     /**
      * Bean to create BasicDataSource.
```
No need for overly verbose Javadocs. Write important things, not names of objects you document, repeated or iterated.

LIttleAncientForestKami commented on robot-db/src/main/java/dbconfiguration/SpringDBConfiguration.java:
```diff
@@ -46,20 +52,30 @@ private Properties hibernateProperties() {
 
+    /**
+     * Entity manager factory bean.
+     * @return - new factory bean of LocalContainerEntityManager
```
As above.

LIttleAncientForestKami commented on robot-db/src/main/java/dbconfiguration/SpringDBConfiguration.java:
```diff
@@ -46,20 +52,30 @@ private Properties hibernateProperties() {
+        adapter.setDatabase(Database.POSTGRESQL);
+        entityManagerFactory.setJpaVendorAdapter(adapter);
+        entityManagerFactory.setDataSource(dataSource());
+        entityManagerFactory.setPackagesToScan("domain");
```
Very good candidate for an external property. If there ever is a time where you want to change packages scanned for beans to something else, it will be easy to do, and may not necessarily require restarting every service.

LIttleAncientForestKami commented on robot-db/src/main/java/dbconfiguration/SpringDBConfiguration.java:
```diff
@@ -46,20 +52,30 @@ private Properties hibernateProperties() {
+    /**
+     * JPA transaction manager.
+     * @return - JpaTransactionManager
+     */
```
As above.

LIttleAncientForestKami commented on robot-db/src/main/java/dbconfiguration/SpringDBConfiguration.java:
```diff
@@ -46,20 +52,30 @@ private Properties hibernateProperties() {
-        HibernateTransactionManager transactionManager = new HibernateTransactionManager();
-        transactionManager.setSessionFactory(sessionFactory);
+    public JpaTransactionManager transactionManager() {
+        JpaTransactionManager transactionManager = new JpaTransactionManager();
```
:+1:

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Author.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.Collection;
+
+/**
+ * Author representation in Database
```
English: either a Database or the Database, but it can't be just "Database". They really notice this stuff.

Nice and short object where Lombok can shine. One thing here that's important is to describe business need for that object. Like:

`Every book has author (one or many). This object represents those authors in the DB.`

Having just "representation of object in DB" is not enough. Business wise this tells folks nothing.

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -0,0 +1,75 @@
+
+
+/**
+ * Book representation in database.
```
As above.

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -0,0 +1,75 @@
+    @Getter
+    @Setter
+    @Column(length = 3)
+    private String currency;
```
I'd note which currency standard we're following. ISO-number goes here or some other code?

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -0,0 +1,75 @@
+    @GeneratedValue
+    @Setter
+    @Getter
+    private long bookID;
```
I'm curious why long?

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -0,0 +1,75 @@
+    @Getter
+    @Setter
+    @Column
+    private String title;
```
How do you deal with length and special diacritical characters?

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -0,0 +1,75 @@
+    @Setter
+    @Column(name = "new_price")
+    private float newPrice;
+
```
Not that I criticize, but putting price details in book entity is an interesting move. What caused it?

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Getter
+    @Setter
+    @Lob
```
A word of justification why LOB would be welcome. Having it in class description is fine.

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Books.java:
```diff
@@ -1,32 +0,0 @@
-import javax.persistence.Id;
-
-/**
- * First idea of entity representing books.
```
I know it's deleted code, but this is actually fairly interesting piece of info in a Javadoc. First idea - nice. :-)

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Category.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Collection;
+
+/**
+ * Category Representation in Database
```
Which tells me NOTHING at all about what this "category" really is and what we're representing here. Useless Javadoc.

LIttleAncientForestKami commented on robot-db/src/main/java/domain/Category.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Getter
+    @Setter
+    @Enumerated(EnumType.STRING)
```
I hope you DO have a proper note on your enumerated type that mentions that sorting and reordering is absolutely forbidden. :imp:

LIttleAncientForestKami commented on robot-db/src/main/java/domain/CategoryName.java:
```diff
@@ -0,0 +1,13 @@
+package domain;
+
+/**
+ * Enum with available book categories
```
Ohh... you DON'T! :P :P :P :cop: :skull: :alien:

LIttleAncientForestKami commented on robot-db/src/main/java/domain/CategoryName.java:
```diff
@@ -0,0 +1,13 @@
+    EDUCATION_AND_SCIENCE,
+    TRAVEL,
+    LIFESTYLE,
+    SEX,
```
I truly wonder why that category is here. Truly puzzling. Utterly. Bewildering. No, really.

LIttleAncientForestKami commented on robot-db/src/main/java/domain/CategoryName.java:
```diff
@@ -0,0 +1,13 @@
+    LIFESTYLE,
+    SEX,
+    MEDICINE,
+    ADVENTURE;
```
Is the order significant? First sex, then medicine, finally adventure? Shouldn't it be the other way around?

LIttleAncientForestKami commented on robot-db/src/main/java/domain/PrintHouse.java:
```diff
@@ -0,0 +1,27 @@
+import javax.persistence.Id;
+
+/**
+ * Simple PrintHouse entity.
```
And now I worry. Simple? As hell.

EDIT: OK, I shouldn't have worried.

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/AuthorInput.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Map;
+
+/**
+ * Class tests if Author with given name is in database if not it put him
```
Is this a test or not?
We shall learn in new episode!

PS. it put -> it puts, English, 3rd person, use `s` letter or I'll send the :cop: your way :p

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/AuthorInput.java:
```diff
@@ -0,0 +1,39 @@
+    @PostConstruct
+    private void init () {
+        authors = new HashMap<>();
+        authorDAO.findAll().forEach(author -> authors.put(author.getName(), author));
```
Potentially dangerous. If someone ever puts `@Cascade` on your author... you are in world of pain, using DAO on `init` like that.

I'm unsure how to advise cause I don't get why this class was created.

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/AuthorInput.java:
```diff
@@ -0,0 +1,39 @@
+        Author author = Author.builder().name(name).build();
+        authors.put(name, author);
+        authorDAO.save(author);
+        return author;
```
No mention whether cache succeeded or failed. Perhaps you don't care, but it's a bad practice. Obviously you call save in only one case here. So, perhaps logging? To have a trace that it happened? Or that it needn't happen?

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/BookInput.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.stream.Collectors;
+
+/**
+ * Class responsible for Inserting Parsed Book into Database
```
DoEs CaSe MaTTeR? iS iT imPorTanT? or not? Now I wonder. In legal documents, now "Inserting Parsed Book into Database" has special meaning. Here?

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/BookInput.java:
```diff
@@ -0,0 +1,62 @@
+     * @param parsedBook - book obtained from Parser.
+     */
+    public void save (ParsedBook parsedBook) {
+        log.info("staring operations on database");
```
Staring at your staring. :eyes: I guess we shall see who will starT to need :eyeglasses: first. With all that staring (:eyes:).

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/BookInput.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    public void save (ParsedBook parsedBook) {
+        log.info("staring operations on database");
+        Category category = categoryInput.saveCategory(parsedBook.getCategory());
```
Try reading that out loud and then tell me it sounds well to your ears, that SAVING a category to DB returns you a category to operate, and that this is a category INPUT operation. What are your DAO doing?

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/BookInput.java:
```diff
@@ -0,0 +1,62 @@
+                .year(parsedBook.getYear())
+                .build();
+        bookDAO.save(book);
+        log.info("saving book" + book);
```
Will there ever be a log message saying "I failed saving the book"? Because for now it seems like no matter whether DAO fails or succeeds, log will say "saving book". Is that :ok:?

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/CategoryInput.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Map;
+
+/**
+ * Class tests if Category with given CategoryName exists in database if not it put him
```
As above..

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/CategoryInput.java:
```diff
@@ -0,0 +1,43 @@
+        categories.put(name, author);
+        categoryDAO.save(author);
+        return author;
+
```
All previous remarks apply. Input, naming, Javadoc, DAO call from Spring Init, etc.

LIttleAncientForestKami commented on robot-db/src/main/java/repositories/ParsedBook.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Class contains all information needed for insert a book into database
+ */
+
```
Cut please.

LIttleAncientForestKami commented:
I'm halfway through, but there's already a high amount of repeat offenders. So please, fix these first and notify me to come back once they're done. For now, :do_not_litter: == no merge.

czyzu0507 committed:
Fixing and Javadoc

clinton56 commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -7,7 +7,11 @@
 
 /**
- * Book representation in database.
+ * {@link Book} is representation of book in a database.
```
This is a very good JavaDoc comparing to that `interface for CRUD Operations on ...`. :+1:

clinton56 commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -72,4 +74,17 @@
     private Category category;
+
+    @Override
+    public String toString () {
```
Anyone using this API will be thankful for `toString()` method. :+1: :ghost:

clinton56 commented on robot-db/src/main/java/domain/Category.java:
```diff
@@ -6,7 +6,9 @@
 
 /**
- * Category Representation in Database
+ * Representation of an {@link CategoryName} a Database.
```
Here probably `in`word  is missing. 
> Representation of an {@link CategoryName} `in` a Database.

clinton56 commented on robot-db/src/main/java/domain/Category.java:
```diff
@@ -6,7 +6,9 @@
- * Category Representation in Database
+ * Representation of an {@link CategoryName} a Database.
+ * Category is mapped {@code @OnoToMany} to {@link Book}
+ * Do not use CascadeType or fetchType.Eager due to performance issue
```
Very nice warning :octocat:  :+1:

clinton56 commented on robot-db/src/main/java/repositories/BookSaver.java:
```diff
@@ -13,36 +13,35 @@
-        log.info("staring operations on database");
-        Category category = categoryInput.saveCategory(parsedBook.getCategory());
-        PrintHouse printHouse = printHouseInput.savePrintHouse(parsedBook.getPrintHouse());
+        Category category = categoryManager.getCategory(parsedBook.getCategory());
```
It would be nice to make `getCategory()` from `CategoryManager` accepting `ParsedBook` object, ex.
before:
> Category category = categoryManager.getCategory(parsedBook.getCategory());
>PrintHouse printHouse = printHouseManager.getPrintHouse(parsedBook.getPrintHouse());

after:

` Category category = categoryManager.getCategory(parsedBook);`
`PrintHouse printHouse = printHouseManager.getPrintHouse(parsedBook);`

:ghost:

clinton56 commented on robot-db/src/main/java/repositories/BookSaver.java:
```diff
@@ -13,36 +13,35 @@
         List<Author> authors = parsedBook.getAuthors()
                 .stream()
-                .map(authorInput::saveAuthor)
+                .map(authorManager::getAuthorForName)
```
:+1:  stream

maciejkocur commented:
:+1:

danone771909 commented:
After corrects it looks very nice, great job :+1:
